hamilton/graph.py

```diff
@@ -150,67 +151,233 @@ def create_graphviz_graph(
-                label = f"Override: {n.name}"
-        is_expand_node = n.node_role == node.NodeType.EXPAND
-        is_collect_node = n.node_role == node.NodeType.COLLECT
+        if n.node_role == node.NodeType.COLLECT:
+            modifier_style = _get_function_modifier_style("collect")
+            node_style.update(**modifier_style)
+            seen_node_types.add("collect")
+
+        if n._tags.get("hamilton.data_saver"):
+            materializer_type = n._tags["hamilton.data_saver.classname"]
+            label = _get_node_label(n, type_=materializer_type)
+            modifier_style = _get_function_modifier_style("materializer")
+            node_style.update(**modifier_style)
+            seen_node_types.add("materializer")
 
-        if is_collect_node or is_expand_node:
-            other_args["peripheries"] = "2"
-        digraph.node(n.name, label=label, **other_args)
+        digraph.node(n.name, label=label, **node_style)
+
+    # create edges
+    input_sets = dict()
+    for n in nodes:
+        input_nodes = set()
 
-    for n in list(nodes):
         for d in n.dependencies:
             if strictly_display_only_nodes_passed_in and d not in nodes:
                 continue
-            if (
-                d not in nodes
-                and d.name in node_modifiers
-                and VisualizationNodeModifiers.IS_USER_INPUT in node_modifiers[d.name]
-            ):
-                digraph.node(d.name, label=f"Input: {d.name}", style="dashed")
-            from_modifiers = node_modifiers.get(d.name, set())
-            to_modifiers = node_modifiers.get(n.name, set())
-            other_args = {}
-            if (
-                VisualizationNodeModifiers.IS_PATH in from_modifiers
-                and VisualizationNodeModifiers.IS_PATH in to_modifiers
-            ):
-                other_args["color"] = PATH_COLOR
-            is_collect_edge = n.node_role == node.NodeType.COLLECT
-            is_expand_edge = d.node_role == node.NodeType.EXPAND
-
-            if is_collect_edge:
-                other_args["dir"] = "both"
-                other_args["arrowtail"] = "crow"
-
-            if is_expand_edge:
-                other_args["dir"] = "both"
-                other_args["arrowhead"] = "crow"
-                other_args["arrowtail"] = "none"
-            digraph.edge(d.name, n.name, **other_args)
+
+            node_type = _get_node_type(d)
+            if node_type == "input":
+                input_nodes.add(d)
+                continue
+
+            edge_style = _get_edge_style(d)
+            digraph.edge(d.name, n.name, **edge_style)
+
+        # skip input node creation
+        if hide_inputs:
+            continue
+
+        if len(input_nodes) > 0:
+            input_node_name = f"{n.name}_inputs"
+
+            # following block is for input node deduplication
+            input_nodes = frozenset(input_nodes)
+            if existing_input_name := input_sets.get(input_nodes):
```

Comment:
Walrus operator doesn't work for 3.7 -- we're going to kill support shortly, but TBD if you want to wait for that..